sdk/eventhub/Azure.Messaging.EventHubs/src/Amqp/AmqpConnectionScope.cs

```diff
@@ -256,6 +259,39 @@ protected AmqpConnectionScope()
             return link;
         }
 
+        /// <summary>
+        ///   Opens an AMQP link for use with producer operations.
+        /// </summary>
+        ///
+        /// <param name="partitionId">The identifier of the Event Hub partition to which the link should be bound; if unbound, <c>null</c>.</param>
+        /// <param name="timeout">The timeout to apply when creating the link.</param>
+        /// <param name="cancellationToken">An optional <see cref="CancellationToken"/> instance to signal the request to cancel the operation.</param>
+        ///
+        /// <returns>A link for use with producer operations.</returns>
+        ///
+        public virtual async Task<SendingAmqpLink> OpenProducerLinkAsync(string partitionId,
+                                                                         TimeSpan timeout,
+                                                                         CancellationToken cancellationToken)
+        {
+            cancellationToken.ThrowIfCancellationRequested<TaskCanceledException>();
+
+            var stopWatch = Stopwatch.StartNew();
+            var path = (string.IsNullOrEmpty(partitionId)) ? EventHubName : string.Format(PartitionProducerPathSuffixMask, EventHubName, partitionId);
+            var producerEndpoint = new Uri(ServiceEndpoint, path);
+
+            var connection = await ActiveConnection.GetOrCreateAsync(timeout).ConfigureAwait(false);
+            cancellationToken.ThrowIfCancellationRequested<TaskCanceledException>();
+
+            var link = await CreateSendingLinkAsync(connection, producerEndpoint, timeout.CalculateRemaining(stopWatch.Elapsed), cancellationToken).ConfigureAwait(false);
+            cancellationToken.ThrowIfCancellationRequested<TaskCanceledException>();
+
+            await OpenAmqpObjectAsync(link, timeout.CalculateRemaining(stopWatch.Elapsed)).ConfigureAwait(false);
+            cancellationToken.ThrowIfCancellationRequested<TaskCanceledException>();
```

Comment:
The link has been opened. Should we still verify the cancellation token at this point?

Reply:
I see where you're coming from.  I'm keeping to the pattern of "check after each external operation."     In this case, we could probably remove it and delegate responsibility to the caller to verify cancellation before using the link.... but I also don't think that checking it here is the wrong thing to do either.   At this point of the code, the AMQP library would likely have been signaled to cancel while opening the link, if it had supported cancellation.

I'm inclined to leave it as-is;  the holder of that token has stated their intent as "stop what you're doing", and I'd like to honor that without taking the time to unwind the call stack and return control to the caller.